From 818c437dcc8413038201824a8bb2e20bcea9121a Mon Sep 17 00:00:00 2001
From: Tim S <stahta01@users.sourceforge.net>
Date: Thu, 3 Sep 2015 17:48:38 -0400
Subject: [PATCH 2/2] * cbp_wx28_32: Added CB custom vars and added lib folder
 to some CB projects. (Thanks stahta01)

Made sure that all the .a files are under lib folder.
Added CB_OBJ, CB_LIB, and CB_BIN CB custom vars.
CB_LIB points to lib folder.
CB_BIN points to devel folder.
CB_OBJ points to .objs folder.
Removed some unneeded search folders.
Removed some unneeded libs.
---
 src/CodeBlocks.cbp                                 | 184 ++++++++++++---------
 src/plugins/contrib/SpellChecker/SpellChecker.cbp  |  23 ++-
 src/plugins/contrib/ToolsPlus/ToolsPlus.cbp        |  14 +-
 src/plugins/contrib/cb_koders/cb_koders.cbp        |  15 +-
 src/plugins/contrib/devpak_plugin/DevPakPlugin.cbp |  19 ++-
 src/plugins/contrib/help_plugin/help-plugin.cbp    |  30 ++--
 src/plugins/contrib/source_exporter/Exporter.cbp   |  30 ++--
 .../contrib/wxContribItems/wxContribItems.cbp      |  38 ++---
 src/plugins/contrib/wxSmith/wxSmith.cbp            |  18 +-
 src/plugins/contrib/wxSmithAui/wxSmithAui.cbp      |  30 ++--
 src/plugins/contrib/wxSmithPlot/wxsSmithPlot.cbp   |  18 +-
 src/tools/cb_share_config/cb_share_config.cbp      |  11 +-
 12 files changed, 245 insertions(+), 185 deletions(-)

diff --git a/src/CodeBlocks.cbp b/src/CodeBlocks.cbp
index 6b8b55a..029ec62 100644
--- a/src/CodeBlocks.cbp
+++ b/src/CodeBlocks.cbp
@@ -7,8 +7,9 @@
 		<Option compiler="gcc" />
 		<Build>
 			<Target title="tinyXML">
-				<Option output="base/tinyxml/txml" prefix_auto="1" extension_auto="1" />
+				<Option output="$(CB_LIB)/txml" prefix_auto="1" extension_auto="1" />
 				<Option working_dir="" />
+				<Option object_output="$(CB_OBJ)" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Compiler>
@@ -17,6 +18,7 @@
 			</Target>
 			<Target title="AutoRevision">
 				<Option output="build_tools/autorevision/autorevision" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)" />
 				<Option type="1" />
 				<Option compiler="gcc" />
 				<Option projectLinkerOptionsRelation="1" />
@@ -26,12 +28,13 @@
 				<Linker>
 					<Add option="-s" />
 					<Add library="txml" />
-					<Add directory="base/tinyxml" />
+					<Add directory="$(CB_LIB)" />
 				</Linker>
 			</Target>
 			<Target title="ConsoleRunner">
 				<Option output="tools/ConsoleRunner/cb_console_runner" prefix_auto="1" extension_auto="1" />
 				<Option working_dir="tools/ConsoleRunner" />
+				<Option object_output="$(CB_OBJ)" />
 				<Option type="1" />
 				<Option compiler="gcc" />
 				<Option projectLinkerOptionsRelation="1" />
@@ -40,8 +43,9 @@
 				</Compiler>
 			</Target>
 			<Target title="Squirrel">
-				<Option output="sdk/scripting/lib/squirrel" prefix_auto="1" extension_auto="1" />
+				<Option output="$(CB_LIB)/squirrel" prefix_auto="1" extension_auto="1" />
 				<Option working_dir="" />
+				<Option object_output="$(CB_OBJ)" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Compiler>
@@ -51,9 +55,10 @@
 				</Compiler>
 			</Target>
 			<Target title="Squirrel std lib">
-				<Option output="sdk/scripting/lib/sqstdlib" prefix_auto="1" extension_auto="1" />
+				<Option output="$(CB_LIB)/sqstdlib" prefix_auto="1" extension_auto="1" />
 				<Option working_dir="" />
-				<Option external_deps="sdk/scripting/lib/libsquirrel.a;" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libsquirrel.a;" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Compiler>
@@ -63,9 +68,10 @@
 				</Compiler>
 			</Target>
 			<Target title="SqPlus">
-				<Option output="sdk/scripting/lib/sqplus" prefix_auto="1" extension_auto="1" />
+				<Option output="$(CB_LIB)/sqplus" prefix_auto="1" extension_auto="1" />
 				<Option working_dir="" />
-				<Option external_deps="sdk/scripting/lib/libsquirrel.a;sdk/scripting/lib/libsqstdlib.a;" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libsquirrel.a;$(CB_LIB)/libsqstdlib.a;" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Compiler>
@@ -75,8 +81,9 @@
 				</Compiler>
 			</Target>
 			<Target title="scintilla">
-				<Option output="devel/wxscintilla_cb" prefix_auto="1" extension_auto="1" />
+				<Option output="$(CB_LIB)/wxscintilla_cb" prefix_auto="1" extension_auto="1" />
 				<Option working_dir="" />
+				<Option object_output="$(CB_OBJ)" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Compiler>
@@ -101,8 +108,9 @@
 				</Linker>
 			</Target>
 			<Target title="wxpropgrid">
-				<Option output="devel/wxpropgrid" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
+				<Option output="$(CB_BIN)/wxpropgrid" imp_lib="$(CB_LIB)/$(TARGET_OUTPUT_BASENAME)" def_file="$(TARGET_OUTPUT_DIR)$(TARGET_OUTPUT_BASENAME)" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option createStaticLib="1" />
@@ -123,9 +131,10 @@
 				</Linker>
 			</Target>
 			<Target title="sdk">
-				<Option output="devel/codeblocks" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="base/tinyxml/libtxml.a;sdk/scripting/lib/libsquirrel.a;sdk/scripting/lib/libsqstdlib.a;sdk/scripting/lib/libsqplus.a;devel/libwxscintilla_cb.a;" />
+				<Option output="$(CB_BIN)/codeblocks" imp_lib="$(CB_LIB)/$(TARGET_OUTPUT_BASENAME)" def_file="$(TARGET_OUTPUT_DIR)$(TARGET_OUTPUT_BASENAME)" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libtxml.a;$(CB_LIB)/libsquirrel.a;$(CB_LIB)/libsqstdlib.a;$(CB_LIB)/libsqplus.a;$(CB_LIB)/libwxscintilla_cb.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --verbose -p debug" />
@@ -163,9 +172,8 @@
 					<Add library="sqstdlib" />
 					<Add library="squirrel" />
 					<Add library="gdi32" />
-					<Add directory="devel" />
-					<Add directory="sdk/scripting/lib" />
-					<Add directory="sdk/propgrid" />
+					<Add directory="$(CB_LIB)" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add before="build_tools\autorevision\autorevision +wx +int +t . include/autorevision.h" />
@@ -176,9 +184,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="src">
-				<Option output="devel/codeblocks" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/codeblocks" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="0" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -217,7 +226,7 @@
 					<Add library="rpcrt4" />
 					<Add library="advapi32" />
 					<Add library="wsock32" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 					<Add directory="exchndl/win32/lib" />
 				</Linker>
 				<ExtraCommands>
@@ -232,9 +241,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="Abbreviations">
-				<Option output="devel/share/CodeBlocks/plugins/abbreviations" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/abbreviations" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -256,7 +266,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -265,9 +275,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="AStyle">
-				<Option output="devel/share/CodeBlocks/plugins/astyle" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/astyle" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -288,7 +299,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -297,9 +308,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="Autosave">
-				<Option output="devel/share/CodeBlocks/plugins/autosave" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/autosave" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -318,7 +330,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -327,8 +339,9 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="Compiler depslib">
-				<Option output="plugins/compilergcc/depslib/depslib" prefix_auto="1" extension_auto="1" />
+				<Option output="$(CB_LIB)/depslib" prefix_auto="1" extension_auto="1" />
 				<Option working_dir="" />
+				<Option object_output="$(CB_OBJ)" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Compiler>
@@ -340,9 +353,10 @@
 				</Linker>
 			</Target>
 			<Target title="Compiler">
-				<Option output="devel/share/CodeBlocks/plugins/compiler" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;plugins/compilergcc/depslib/libdepslib.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/compiler" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;$(CB_LIB)/libdepslib.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -366,8 +380,8 @@
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
 					<Add library="depslib" />
 					<Add library="wxpropgrid" />
-					<Add directory="devel" />
-					<Add directory="plugins/compilergcc/depslib" />
+					<Add directory="$(CB_LIB)" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -379,9 +393,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="Debugger">
-				<Option output="devel/share/CodeBlocks/plugins/debugger" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/debugger" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -402,7 +417,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -411,9 +426,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="Code-completion">
-				<Option output="devel/share/CodeBlocks/plugins/codecompletion" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/codecompletion" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -438,7 +454,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -447,9 +463,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="Class wizard">
-				<Option output="devel/share/CodeBlocks/plugins/classwizard" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/classwizard" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -468,7 +485,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -477,9 +494,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="Default MIME handler">
-				<Option output="devel/share/CodeBlocks/plugins/defaultmimehandler" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/defaultmimehandler" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -498,7 +516,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -507,9 +525,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="Occurrences Highlighting">
-				<Option output="devel/share/CodeBlocks/plugins/OccurrencesHighlighting" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/OccurrencesHighlighting" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -528,7 +547,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -537,9 +556,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="Open files list">
-				<Option output="devel/share/CodeBlocks/plugins/openfileslist" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/openfileslist" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -558,7 +578,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -567,9 +587,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="Projects-workspaces importer">
-				<Option output="devel/share/CodeBlocks/plugins/projectsimporter" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/projectsimporter" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -594,7 +615,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -603,9 +624,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="Scripted wizard">
-				<Option output="devel/share/CodeBlocks/plugins/scriptedwizard" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/scriptedwizard" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -626,7 +648,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -635,9 +657,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="To-do">
-				<Option output="devel/share/CodeBlocks/plugins/todo" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/todo" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -662,7 +685,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -671,9 +694,10 @@
 				</ExtraCommands>
 			</Target>
 			<Target title="XP look &amp; feel">
-				<Option output="devel/share/CodeBlocks/plugins/xpmanifest" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="devel" />
-				<Option external_deps="devel/libcodeblocks.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/xpmanifest" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-dde --no-check-associations --multiple-instance --no-splash-screen --verbose -p debug" />
@@ -692,7 +716,7 @@
 					<Add option="-mthreads" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="devel" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist devel\share\CodeBlocks mkdir devel\share\CodeBlocks" />
@@ -701,6 +725,9 @@
 				</ExtraCommands>
 			</Target>
 			<Environment>
+				<Variable name="CB_BIN" value="devel" />
+				<Variable name="CB_LIB" value="lib" />
+				<Variable name="CB_OBJ" value=".objs" />
 				<Variable name="WX_CFG" value="" />
 				<Variable name="WX_SUFFIX" value="u" />
 				<Variable name="WX_VERSION" value="28" />
@@ -738,7 +765,6 @@
 		</ResourceCompiler>
 		<Linker>
 			<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-			<Add directory="base/tinyxml" />
 			<Add directory="$(#WX.lib)/gcc_dll$(WX_CFG)" />
 		</Linker>
 		<Unit filename="base/tinyxml/tinystr.cpp">
diff --git a/src/plugins/contrib/SpellChecker/SpellChecker.cbp b/src/plugins/contrib/SpellChecker/SpellChecker.cbp
index 241ae47..12e85eb 100644
--- a/src/plugins/contrib/SpellChecker/SpellChecker.cbp
+++ b/src/plugins/contrib/SpellChecker/SpellChecker.cbp
@@ -8,9 +8,9 @@
 		<Option compiler="gcc" />
 		<Build>
 			<Target title="hunspell">
-				<Option output="../../../devel/hunspell" prefix_auto="1" extension_auto="1" />
+				<Option output="$(CB_LIB)/plugins/contrib/SpellChecker/wxSpellChecker/hunspell" prefix_auto="1" extension_auto="1" />
 				<Option working_dir="" />
-				<Option object_output="../../../.objs/plugins/contrib/SpellChecker/hunspell" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/SpellChecker/hunspell" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Option createDefFile="1" />
@@ -34,9 +34,9 @@
 				</Linker>
 			</Target>
 			<Target title="wxSpellChecker">
-				<Option output="../../../devel/wxSpellChecker" prefix_auto="1" extension_auto="1" />
+				<Option output="$(CB_LIB)/plugins/contrib/SpellChecker/wxSpellChecker/wxSpellChecker" prefix_auto="1" extension_auto="1" />
 				<Option working_dir="" />
-				<Option object_output="../../../.objs/plugins/contrib/SpellChecker/wxSpellChecker" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/SpellChecker/wxSpellChecker" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Option createDefFile="1" />
@@ -56,17 +56,16 @@
 				<Linker>
 					<Add library="hunspell" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="../../../devel" />
 					<Add directory="$(#wx.lib)/gcc_dll$(WX_CFG)" />
 				</Linker>
 			</Target>
 			<Target title="plugin">
-				<Option output="../../../devel/share/CodeBlocks/plugins/SpellChecker" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/SpellChecker" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/SpellChecker" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/SpellChecker" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --no-check-associations -ns -nd --multiple-instance -p debug" />
-				<Option host_application="../../../devel/codeblocks.exe" />
+				<Option host_application="$(CB_BIN)/codeblocks.exe" />
 				<Option run_host_application_in_terminal="0" />
 				<Compiler>
 					<Add option="$(#CB_RELEASE_TYPE)" />
@@ -91,8 +90,9 @@
 					<Add library="hunspell" />
 					<Add library="codeblocks" />
 					<Add library="wxmsw$(WX_VERSION)$(WX_SUFFIX)" />
-					<Add directory="../../../devel" />
 					<Add directory="$(#wx.lib)/gcc_dll$(WX_CFG)" />
+					<Add directory="$(CB_LIB)/plugins/contrib/SpellChecker/wxSpellChecker" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
 					<Add after="cmd /c if not exist ..\..\..\devel\share\CodeBlocks\SpellChecker mkdir ..\..\..\devel\share\CodeBlocks\SpellChecker" />
@@ -115,6 +115,11 @@
 					<Variable name="WX_VERSION" value="28" />
 				</Environment>
 			</Target>
+			<Environment>
+				<Variable name="CB_BIN" value="..\..\..\devel" />
+				<Variable name="CB_LIB" value="..\..\..\lib" />
+				<Variable name="CB_OBJ" value="..\..\..\.objs" />
+			</Environment>
 		</Build>
 		<VirtualTargets>
 			<Add alias="All" targets="hunspell;wxSpellChecker;plugin;" />
diff --git a/src/plugins/contrib/ToolsPlus/ToolsPlus.cbp b/src/plugins/contrib/ToolsPlus/ToolsPlus.cbp
index 002d635..0c60d91 100644
--- a/src/plugins/contrib/ToolsPlus/ToolsPlus.cbp
+++ b/src/plugins/contrib/ToolsPlus/ToolsPlus.cbp
@@ -6,26 +6,28 @@
 		<Option compiler="gcc" />
 		<Build>
 			<Target title="default">
-				<Option output="../../../devel/share/codeblocks/plugins/ToolsPlus" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/ToolsPlus" />
+				<Option output="$(CB_BIN)/share/codeblocks/plugins/ToolsPlus" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/ToolsPlus" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --multiple-instance -na -ns -nd -p debug" />
-				<Option host_application="../../../devel/codeblocks.exe" />
+				<Option host_application="$(CB_BIN)/codeblocks.exe" />
 				<Option run_host_application_in_terminal="0" />
 				<Compiler>
 					<Add option="$(#CB_RELEASE_TYPE)" />
 				</Compiler>
 				<Linker>
 					<Add library="codeblocks" />
-					<Add library="wxscintilla_cb" />
 					<Add library="wxmsw28$(WX_SUFFIX)" />
 					<Add library="gdi32" />
-					<Add directory="../../../devel" />
 					<Add directory="$(#wx.lib)/gcc_dll$(WX_CFG)" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 			</Target>
 			<Environment>
+				<Variable name="CB_BIN" value="..\..\..\devel" />
+				<Variable name="CB_LIB" value="..\..\..\lib" />
+				<Variable name="CB_OBJ" value="..\..\..\.objs" />
 				<Variable name="WX_CFG" value="" />
 				<Variable name="WX_SUFFIX" value="u" />
 			</Environment>
@@ -63,7 +65,7 @@
 			<Add option="-mthreads" />
 		</Linker>
 		<ExtraCommands>
-			<Add after="zip -j9 ..\..\..\devel\share\codeblocks\ToolsPlus.zip manifest.xml" />
+			<Add after="zip -j9 $(CB_BIN)\share\codeblocks\ToolsPlus.zip manifest.xml" />
 			<Add after="update.bat" />
 			<Mode after="always" />
 		</ExtraCommands>
diff --git a/src/plugins/contrib/cb_koders/cb_koders.cbp b/src/plugins/contrib/cb_koders/cb_koders.cbp
index ef561b0..f60d774 100644
--- a/src/plugins/contrib/cb_koders/cb_koders.cbp
+++ b/src/plugins/contrib/cb_koders/cb_koders.cbp
@@ -8,17 +8,20 @@
 		<Option compiler="gcc" />
 		<Build>
 			<Target title="default">
-				<Option output="../../../devel/share/CodeBlocks/plugins/cb_koders.dll" prefix_auto="0" extension_auto="0" />
-				<Option object_output="../../../.objs/plugins/contrib/cb_koders" />
-				<Option external_deps="../../../devel/libcodeblocks.a;../../../devel/libwxscintilla_cb.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/cb_koders.dll" prefix_auto="0" extension_auto="0" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/cb_koders" />
+				<Option external_deps="$(CB_LIB)/libcodeblocks.a;$(CB_LIB)/libwxscintilla_cb.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --multiple-instance -na -ns -nd -p debug" />
-				<Option host_application="../../../devel/codeblocks.exe" />
+				<Option host_application="$(CB_BIN)/codeblocks.exe" />
 				<Option run_host_application_in_terminal="0" />
 				<Option projectResourceIncludeDirsRelation="0" />
 			</Target>
 			<Environment>
+				<Variable name="CB_BIN" value="..\..\..\devel" />
+				<Variable name="CB_LIB" value="..\..\..\lib" />
+				<Variable name="CB_OBJ" value="..\..\..\.objs" />
 				<Variable name="WX_CFG" value="" />
 				<Variable name="WX_SUFFIX" value="u" />
 			</Environment>
@@ -56,10 +59,10 @@
 			<Add library="wxmsw28$(WX_SUFFIX)" />
 			<Add library="codeblocks" />
 			<Add directory="$(#WX.lib)/gcc_dll$(WX_CFG)" />
-			<Add directory="../../../devel" />
+			<Add directory="$(CB_LIB)" />
 		</Linker>
 		<ExtraCommands>
-			<Add after="zip -j9 ..\..\..\devel\share\CodeBlocks\cb_koders.zip manifest.xml" />
+			<Add after="zip -j9 $(CB_BIN)\share\CodeBlocks\cb_koders.zip manifest.xml" />
 			<Mode after="always" />
 		</ExtraCommands>
 		<Unit filename="cb_koders.cpp" />
diff --git a/src/plugins/contrib/devpak_plugin/DevPakPlugin.cbp b/src/plugins/contrib/devpak_plugin/DevPakPlugin.cbp
index a691278..6254584 100644
--- a/src/plugins/contrib/devpak_plugin/DevPakPlugin.cbp
+++ b/src/plugins/contrib/devpak_plugin/DevPakPlugin.cbp
@@ -8,9 +8,9 @@
 		<Option compiler="gcc" />
 		<Build>
 			<Target title="libbz2">
-				<Option output="bzip2/libbz2.a" prefix_auto="0" extension_auto="0" />
+				<Option output="$(CB_LIB)/plugins/contrib/devpak_plugin/libbz2.a" prefix_auto="0" extension_auto="0" />
 				<Option working_dir="" />
-				<Option object_output="../../../.objs/plugins/contrib/devpak_plugin" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/devpak_plugin" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Compiler>
@@ -21,12 +21,12 @@
 				</Linker>
 			</Target>
 			<Target title="default">
-				<Option output="../../../devel/share/CodeBlocks/plugins/devpakupdater.dll" prefix_auto="0" extension_auto="0" />
-				<Option object_output="../../../.objs/plugins/contrib/devpak_plugin" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/devpakupdater.dll" prefix_auto="0" extension_auto="0" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/devpak_plugin" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --multiple-instance -na -ns -nd -p debug" />
-				<Option host_application="../../../devel/codeblocks.exe" />
+				<Option host_application="$(CB_BIN)/codeblocks.exe" />
 				<Option run_host_application_in_terminal="0" />
 				<Option projectResourceIncludeDirsRelation="0" />
 				<Compiler>
@@ -48,14 +48,18 @@
 				<Linker>
 					<Add option="-mthreads" />
 					<Add library="bz2" />
-					<Add directory="../../../devel" />
+					<Add directory="$(CB_LIB)" />
+					<Add directory="$(CB_LIB)/plugins/contrib/devpak_plugin" />
 				</Linker>
 				<ExtraCommands>
-					<Add after="zip ..\..\..\devel\share\CodeBlocks\devpakupdater.zip manifest.xml *.xrc" />
+					<Add after="zip $(CB_BIN)\share\CodeBlocks\devpakupdater.zip manifest.xml *.xrc" />
 					<Mode after="always" />
 				</ExtraCommands>
 			</Target>
 			<Environment>
+				<Variable name="CB_BIN" value="..\..\..\devel" />
+				<Variable name="CB_LIB" value="..\..\..\lib" />
+				<Variable name="CB_OBJ" value="..\..\..\.objs" />
 				<Variable name="WX_CFG" value="" />
 				<Variable name="WX_SUFFIX" value="u" />
 			</Environment>
@@ -75,7 +79,6 @@
 			<Add option="-Wl,--enable-auto-import" />
 			<Add library="codeblocks" />
 			<Add library="wxmsw28$(WX_SUFFIX)" />
-			<Add directory="bzip2" />
 			<Add directory="$(#WX.lib)/gcc_dll$(WX_CFG)" />
 		</Linker>
 		<Unit filename="bzip2/blocksort.c">
diff --git a/src/plugins/contrib/help_plugin/help-plugin.cbp b/src/plugins/contrib/help_plugin/help-plugin.cbp
index 7015570..3b2cbe2 100644
--- a/src/plugins/contrib/help_plugin/help-plugin.cbp
+++ b/src/plugins/contrib/help_plugin/help-plugin.cbp
@@ -8,9 +8,9 @@
 		<Option compiler="gcc" />
 		<Build>
 			<Target title="bzip2">
-				<Option output="bzip2/libbzip2.a" prefix_auto="0" extension_auto="0" />
+				<Option output="$(CB_LIB)/plugins/contrib/help_plugin/libbzip2.a" prefix_auto="0" extension_auto="0" />
 				<Option working_dir="" />
-				<Option object_output="../../../.objs/plugins/contrib/help_plugin/bzip2" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/help_plugin/bzip2" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Option createDefFile="1" />
@@ -22,9 +22,9 @@
 				</Linker>
 			</Target>
 			<Target title="zlib">
-				<Option output="zlib/libz.a" prefix_auto="0" extension_auto="0" />
+				<Option output="$(CB_LIB)/plugins/contrib/help_plugin/libz.a" prefix_auto="0" extension_auto="0" />
 				<Option working_dir="" />
-				<Option object_output="../../../.objs/plugins/contrib/help_plugin/zlib" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/help_plugin/zlib" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Option createDefFile="1" />
@@ -36,12 +36,12 @@
 				</Linker>
 			</Target>
 			<Target title="help_plugin">
-				<Option output="../../../devel/share/CodeBlocks/plugins/help_plugin.dll" prefix_auto="0" extension_auto="0" />
-				<Option object_output="../../../.objs/plugins/contrib/help_plugin" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/help_plugin.dll" prefix_auto="0" extension_auto="0" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/help_plugin" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --multiple-instance -na -ns -nd -p debug" />
-				<Option host_application="../../../devel/codeblocks.exe" />
+				<Option host_application="$(CB_BIN)/codeblocks.exe" />
 				<Option run_host_application_in_terminal="0" />
 				<Compiler>
 					<Add option="-Wall" />
@@ -64,19 +64,24 @@
 				</Compiler>
 				<Linker>
 					<Add option="-mthreads" />
-					<Add library="bzip2\libbzip2.a" />
-					<Add library="zlib\libz.a" />
+					<Add library="libbzip2.a" />
+					<Add library="libz.a" />
+					<Add directory="$(CB_LIB)/plugins/contrib/help_plugin" />
+					<Add directory="$(CB_LIB)" />
 				</Linker>
 				<ExtraCommands>
-					<Add after="zip -j9 ..\..\..\devel\share\CodeBlocks\help_plugin.zip manifest.xml *.xrc *.png" />
-					<Add after="cmd /c if not exist ..\..\..\devel\share\CodeBlocks\docs mkdir ..\..\..\devel\share\CodeBlocks\docs" />
-					<Add after="xcopy /Y index.ini ..\..\..\devel\share\CodeBlocks\docs" />
+					<Add after="zip -j9 $(CB_BIN)\share\CodeBlocks\help_plugin.zip manifest.xml *.xrc *.png" />
+					<Add after="cmd /c if not exist $(CB_BIN)\share\CodeBlocks\docs mkdir $(CB_BIN)\share\CodeBlocks\docs" />
+					<Add after="xcopy /Y index.ini $(CB_BIN)\share\CodeBlocks\docs" />
 					<Add after="cmd /c if not exist ..\..\..\output\share\CodeBlocks\docs mkdir ..\..\..\output\share\CodeBlocks\docs" />
 					<Add after="xcopy /Y index.ini ..\..\..\output\share\CodeBlocks\docs" />
 					<Mode after="always" />
 				</ExtraCommands>
 			</Target>
 			<Environment>
+				<Variable name="CB_BIN" value="..\..\..\devel" />
+				<Variable name="CB_LIB" value="..\..\..\lib" />
+				<Variable name="CB_OBJ" value="..\..\..\.objs" />
 				<Variable name="WX_CFG" value="" />
 				<Variable name="WX_SUFFIX" value="u" />
 			</Environment>
@@ -102,7 +107,6 @@
 			<Add option="-Wl,--enable-auto-import" />
 			<Add library="wxmsw28$(WX_SUFFIX)" />
 			<Add library="codeblocks" />
-			<Add directory="../../../devel" />
 			<Add directory="$(#WX.lib)/gcc_dll$(WX_CFG)" />
 		</Linker>
 		<Unit filename="HelpConfigDialog.cpp">
diff --git a/src/plugins/contrib/source_exporter/Exporter.cbp b/src/plugins/contrib/source_exporter/Exporter.cbp
index 7d2dfe9..108dd78 100644
--- a/src/plugins/contrib/source_exporter/Exporter.cbp
+++ b/src/plugins/contrib/source_exporter/Exporter.cbp
@@ -8,9 +8,9 @@
 		<Option compiler="gcc" />
 		<Build>
 			<Target title="wxPdfDocument">
-				<Option output="wxPdfDocument/lib/wxPdfDocument" prefix_auto="1" extension_auto="1" />
+				<Option output="$(CB_LIB)/plugins/contrib/source_exporter/wxPdfDocument" prefix_auto="1" extension_auto="1" />
 				<Option working_dir="" />
-				<Option object_output="../../../.objs/plugins/contrib/source_exporter" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/source_exporter" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Option projectCompilerOptionsRelation="2" />
@@ -22,47 +22,44 @@
 			<Target title="makefont">
 				<Option output="wxPdfDocument/makefont/makefont" prefix_auto="0" extension_auto="1" />
 				<Option working_dir="wxPdfDocument/makefont" />
-				<Option object_output="../../../.objs/plugins/contrib/source_exporter" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/source_exporter" />
 				<Option type="1" />
 				<Option compiler="gcc" />
 				<Linker>
 					<Add library="wxPdfDocument" />
 					<Add library="wxmsw28$(WX_SUFFIX)" />
-					<Add directory="../../../devel" />
 				</Linker>
 			</Target>
 			<Target title="minimal">
 				<Option output="wxPdfDocument/samples/minimal/minimal" prefix_auto="0" extension_auto="1" />
 				<Option working_dir="wxPdfDocument/samples/minimal" />
-				<Option object_output="../../../.objs/plugins/contrib/source_exporter" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/source_exporter" />
 				<Option type="1" />
 				<Option compiler="gcc" />
 				<Linker>
 					<Add library="wxPdfDocument" />
 					<Add library="wxmsw28$(WX_SUFFIX)" />
-					<Add directory="../../../devel" />
 				</Linker>
 			</Target>
 			<Target title="pdfdc">
 				<Option output="wxPdfDocument/samples/pdfdcl/pdfdc" prefix_auto="0" extension_auto="1" />
 				<Option working_dir="wxPdfDocument/samples/pdfdc" />
-				<Option object_output="../../../.objs/plugins/contrib/source_exporter" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/source_exporter" />
 				<Option type="1" />
 				<Option compiler="gcc" />
 				<Linker>
 					<Add library="wxPdfDocument" />
 					<Add library="wxmsw28$(WX_SUFFIX)" />
-					<Add directory="../../../devel" />
 				</Linker>
 			</Target>
 			<Target title="default">
-				<Option output="../../../devel/share/CodeBlocks/plugins/Exporter" prefix_auto="0" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/source_exporter" />
-				<Option external_deps="wxPdfDocument/lib/libwxPdfDocument.a;" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/Exporter" prefix_auto="0" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/source_exporter" />
+				<Option external_deps="$(CB_LIB)/plugins/contrib/source_exporter/libwxPdfDocument.a;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --multiple-instance -na -ns -nd -p debug" />
-				<Option host_application="../../../devel/codeblocks.exe" />
+				<Option host_application="$(CB_BIN)/codeblocks.exe" />
 				<Option run_host_application_in_terminal="0" />
 				<Option projectLinkerOptionsRelation="2" />
 				<Compiler>
@@ -77,15 +74,18 @@
 					<Add library="wxPdfDocument" />
 					<Add library="wxmsw28$(WX_SUFFIX)" />
 					<Add library="gdi32" />
-					<Add directory="../../../devel" />
-					<Add directory="wxPdfDocument/lib" />
+					<Add directory="$(CB_LIB)/plugins/contrib/source_exporter" />
+					<Add directory="$(CB_LIB)" />
 				</Linker>
 				<ExtraCommands>
-					<Add after="zip -j9 ..\..\..\devel\share\CodeBlocks\Exporter.zip manifest.xml" />
+					<Add after="zip -j9 $(CB_BIN)\share\CodeBlocks\Exporter.zip manifest.xml" />
 					<Mode after="always" />
 				</ExtraCommands>
 			</Target>
 			<Environment>
+				<Variable name="CB_BIN" value="..\..\..\devel" />
+				<Variable name="CB_LIB" value="..\..\..\lib" />
+				<Variable name="CB_OBJ" value="..\..\..\.objs" />
 				<Variable name="WX_CFG" value="" />
 				<Variable name="WX_SUFFIX" value="u" />
 			</Environment>
diff --git a/src/plugins/contrib/wxContribItems/wxContribItems.cbp b/src/plugins/contrib/wxContribItems/wxContribItems.cbp
index 8d732db..b0143ae 100644
--- a/src/plugins/contrib/wxContribItems/wxContribItems.cbp
+++ b/src/plugins/contrib/wxContribItems/wxContribItems.cbp
@@ -9,8 +9,8 @@
 		<Build>
 			<Target title="wxFlatNoteBook">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/wxflatnotebook" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/wxContribItems/wxFlatNotebook" />
+				<Option output="$(CB_BIN)/wxflatnotebook" imp_lib="$(CB_LIB)/$(TARGET_OUTPUT_BASENAME)" def_file="$(TARGET_OUTPUT_DIR)$(TARGET_OUTPUT_BASENAME)" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxContribItems/wxFlatNotebook" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option createStaticLib="1" />
@@ -21,8 +21,8 @@
 			</Target>
 			<Target title="wxChartCtrl">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/wxchartctrl" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/wxContribItems/wxChartCtrl" />
+				<Option output="$(CB_BIN)/wxchartctrl" imp_lib="$(CB_LIB)/$(TARGET_OUTPUT_BASENAME)" def_file="$(TARGET_OUTPUT_DIR)$(TARGET_OUTPUT_BASENAME)" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxContribItems/wxChartCtrl" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option createStaticLib="1" />
@@ -33,8 +33,8 @@
 			</Target>
 			<Target title="wxCustomButton">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/wxcustombutton" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/wxContribItems/wxthings" />
+				<Option output="$(CB_BIN)/wxcustombutton" imp_lib="$(CB_LIB)/$(TARGET_OUTPUT_BASENAME)" def_file="$(TARGET_OUTPUT_DIR)$(TARGET_OUTPUT_BASENAME)" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxContribItems/wxthings" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option createStaticLib="1" />
@@ -45,8 +45,8 @@
 			</Target>
 			<Target title="wxImagePanel">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/wximagepanel" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/wxContribItems/wxImagePanel" />
+				<Option output="$(CB_BIN)/wximagepanel" imp_lib="$(CB_LIB)/$(TARGET_OUTPUT_BASENAME)" def_file="$(TARGET_OUTPUT_DIR)$(TARGET_OUTPUT_BASENAME)" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxContribItems/wxImagePanel" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option createStaticLib="1" />
@@ -59,8 +59,8 @@
 			</Target>
 			<Target title="wxSpeedButton">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/wxspeedbutton" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/wxContribItems/wxSpeedButton" />
+				<Option output="$(CB_BIN)/wxspeedbutton" imp_lib="$(CB_LIB)/$(TARGET_OUTPUT_BASENAME)" def_file="$(TARGET_OUTPUT_DIR)$(TARGET_OUTPUT_BASENAME)" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxContribItems/wxSpeedButton" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option createStaticLib="1" />
@@ -70,8 +70,8 @@
 			</Target>
 			<Target title="wxKWIC">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/wxkwic" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/wxContribItems/KWIC" />
+				<Option output="$(CB_BIN)/wxkwic" imp_lib="$(CB_LIB)/$(TARGET_OUTPUT_BASENAME)" def_file="$(TARGET_OUTPUT_DIR)$(TARGET_OUTPUT_BASENAME)" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxContribItems/KWIC" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option createStaticLib="1" />
@@ -84,8 +84,8 @@
 			</Target>
 			<Target title="wxTreeList">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/wxtreelist" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/wxContribItems/wxtreelist" />
+				<Option output="$(CB_BIN)/wxtreelist" imp_lib="$(CB_LIB)/$(TARGET_OUTPUT_BASENAME)" def_file="$(TARGET_OUTPUT_DIR)$(TARGET_OUTPUT_BASENAME)" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxContribItems/wxtreelist" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option createStaticLib="1" />
@@ -95,8 +95,8 @@
 			</Target>
 			<Target title="wxLED">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/wxled" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/wxContribItems/wxled" />
+				<Option output="$(CB_BIN)/wxled" imp_lib="$(CB_LIB)/$(TARGET_OUTPUT_BASENAME)" def_file="$(TARGET_OUTPUT_DIR)$(TARGET_OUTPUT_BASENAME)" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxContribItems/wxled" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option createStaticLib="1" />
@@ -111,6 +111,9 @@
 				</Linker>
 			</Target>
 			<Environment>
+				<Variable name="CB_BIN" value="..\..\..\devel" />
+				<Variable name="CB_LIB" value="..\..\..\lib" />
+				<Variable name="CB_OBJ" value="..\..\..\.objs" />
 				<Variable name="WX_CFG" value="" />
 				<Variable name="WX_SUFFIX" value="u" />
 			</Environment>
@@ -131,14 +134,11 @@
 			<Add option="-DwxUSE_UNICODE" />
 			<Add directory="$(#wx.include)" />
 			<Add directory="$(#wx.lib)/gcc_dll$(WX_CFG)/msw$(WX_SUFFIX)" />
-			<Add directory="wxFlatNotebook/include" />
 		</Compiler>
 		<Linker>
 			<Add option="-Wl,--enable-auto-import" />
 			<Add option="-mthreads" />
-			<Add library="wxpropgrid" />
 			<Add library="wxmsw28$(WX_SUFFIX)" />
-			<Add directory="../../../devel" />
 			<Add directory="$(#wx.lib)/gcc_dll$(WX_CFG)" />
 		</Linker>
 		<Unit filename="KWIC/include/wx/KWIC/AngularMeter.h">
diff --git a/src/plugins/contrib/wxSmith/wxSmith.cbp b/src/plugins/contrib/wxSmith/wxSmith.cbp
index f0e656c..1f222a4 100644
--- a/src/plugins/contrib/wxSmith/wxSmith.cbp
+++ b/src/plugins/contrib/wxSmith/wxSmith.cbp
@@ -9,27 +9,30 @@
 		<Build>
 			<Target title="wxSmithLib">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/wxsmithlib" prefix_auto="1" extension_auto="1" />
-				<Option working_dir="../../../devel" />
-				<Option object_output="../../../.objs/plugins/contrib/wxSmithLib" />
+				<Option output="$(CB_BIN)/wxsmithlib" imp_lib="$(CB_LIB)/$(TARGET_OUTPUT_BASENAME)" def_file="$(TARGET_OUTPUT_DIR)$(TARGET_OUTPUT_BASENAME)" prefix_auto="1" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxSmithLib" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option createStaticLib="1" />
 			</Target>
 			<Target title="wxSmith">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/share/CodeBlocks/plugins/wxsmith" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/wxSmith" />
+				<Option output="$(CB_BIN)/share/CodeBlocks/plugins/wxsmith" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxSmith" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --multiple-instance -na -ns -nd -p debug" />
-				<Option host_application="../../../devel/codeblocks.exe" />
+				<Option host_application="$(CB_BIN)/codeblocks.exe" />
 				<Option run_host_application_in_terminal="0" />
 				<Linker>
 					<Add library="wxsmithlib" />
 				</Linker>
 			</Target>
 			<Environment>
+				<Variable name="CB_BIN" value="..\..\..\devel" />
+				<Variable name="CB_LIB" value="..\..\..\lib" />
+				<Variable name="CB_OBJ" value="..\..\..\.objs" />
 				<Variable name="WX_CFG" value="" />
 				<Variable name="WX_SUFFIX" value="u" />
 			</Environment>
@@ -84,8 +87,7 @@
 			<Add library="codeblocks" />
 			<Add directory="$(#WX.lib)/gcc_dll$(WX_CFG)" />
 			<Add directory="$(#WX.lib)/gcc_dll$(WX_CFG)/msw$(WX_SUFFIX)" />
-			<Add directory="../../../devel" />
-			<Add directory="../../../base/tinyxml" />
+			<Add directory="$(CB_BIN)" />
 		</Linker>
 		<ExtraCommands>
 			<Add after="update.bat" />
diff --git a/src/plugins/contrib/wxSmithAui/wxSmithAui.cbp b/src/plugins/contrib/wxSmithAui/wxSmithAui.cbp
index 803fe8e..97139b3 100644
--- a/src/plugins/contrib/wxSmithAui/wxSmithAui.cbp
+++ b/src/plugins/contrib/wxSmithAui/wxSmithAui.cbp
@@ -8,13 +8,13 @@
 		<Build>
 			<Target title="Unix">
 				<Option platforms="Unix;Mac;" />
-				<Option output="../../../devel/share/codeblocks/plugins/libwxSmithAui" prefix_auto="0" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/wxSmithAui" />
-				<Option external_deps="../../../devel/libwxsmithlib.so;" />
+				<Option output="$(CB_BIN)/share/codeblocks/plugins/libwxSmithAui" prefix_auto="0" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxSmithAui" />
+				<Option external_deps="$(CB_BIN)/libwxsmithlib.so;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --multiple-instance -ns -ni -p debug" />
-				<Option host_application="../../../devel/codeblocks" />
+				<Option host_application="$(CB_BIN)/codeblocks" />
 				<Option run_host_application_in_terminal="0" />
 				<Compiler>
 					<Add option="-ansi" />
@@ -26,21 +26,26 @@
 				<Linker>
 					<Add option="`wx-config --version=2.8 --libs`" />
 					<Add option="-Wl,--no-undefined" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
-					<Add after="zip -j9 ../../../devel/share/codeblocks/wxSmithAui.zip manifest.xml" />
+					<Add after="zip -j9 $(CB_BIN)/share/codeblocks/wxSmithAui.zip manifest.xml" />
 					<Mode after="always" />
 				</ExtraCommands>
+				<Environment>
+					<Variable name="CB_BIN" value="../../../devel" />
+					<Variable name="CB_OBJ" value="../../../.objs" />
+				</Environment>
 			</Target>
 			<Target title="Windows">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/share/codeblocks/plugins/wxSmithAui" prefix_auto="1" extension_auto="1" />
-				<Option object_output="../../../.objs/plugins/contrib/wxSmithAui" />
-				<Option external_deps="../../../devel/wxsmithlib.dll;" />
+				<Option output="$(CB_BIN)/share/codeblocks/plugins/wxSmithAui" prefix_auto="1" extension_auto="1" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxSmithAui" />
+				<Option external_deps="$(CB_BIN)/wxsmithlib.dll;" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log --multiple-instance -na -ns -nd -p debug" />
-				<Option host_application="../../../devel/codeblocks.exe" />
+				<Option host_application="$(CB_BIN)/codeblocks.exe" />
 				<Option run_host_application_in_terminal="0" />
 				<Compiler>
 					<Add option="$(#CB_RELEASE_TYPE)" />
@@ -62,12 +67,16 @@
 					<Add option="-mthreads" />
 					<Add library="wxmsw28$(WX_SUFFIX)" />
 					<Add directory="$(#wx.lib)/gcc_dll$(WX_CFG)" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
-					<Add after="zip -j9 ..\..\..\devel\share\CodeBlocks\wxSmithAui.zip manifest.xml" />
+					<Add after="zip -j9 $(CB_BIN)\share\CodeBlocks\wxSmithAui.zip manifest.xml" />
 					<Mode after="always" />
 				</ExtraCommands>
 				<Environment>
+					<Variable name="CB_BIN" value="../../../devel" />
+					<Variable name="CB_LIB" value="../../../lib" />
+					<Variable name="CB_OBJ" value="../../../.objs" />
 					<Variable name="WX_CFG" value="" />
 					<Variable name="WX_SUFFIX" value="u" />
 				</Environment>
@@ -93,7 +102,6 @@
 			<Add library="codeblocks" />
 			<Add library="wxsmithlib" />
 			<Add library="wxpropgrid" />
-			<Add directory="../../../devel" />
 		</Linker>
 		<Unit filename="manifest.xml" />
 		<Unit filename="wxAuiManager/wxSmithAuiManager.h" />
diff --git a/src/plugins/contrib/wxSmithPlot/wxsSmithPlot.cbp b/src/plugins/contrib/wxSmithPlot/wxsSmithPlot.cbp
index 89be967..e32c0dc 100644
--- a/src/plugins/contrib/wxSmithPlot/wxsSmithPlot.cbp
+++ b/src/plugins/contrib/wxSmithPlot/wxsSmithPlot.cbp
@@ -7,18 +7,18 @@
 		<Option compiler="gcc" />
 		<Build>
 			<Target title="MathPlot">
-				<Option output="../../../devel/MathPlot" prefix_auto="1" extension_auto="1" />
+				<Option output="$(CB_LIB)/plugins/contrib/wxSmithPlot/MathPlot" prefix_auto="1" extension_auto="1" />
 				<Option working_dir="" />
-				<Option object_output="../../../.objs/plugins/contrib/wxSmithPlot/MathPlot" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxSmithPlot/MathPlot" />
 				<Option type="2" />
 				<Option compiler="gcc" />
 				<Option createDefFile="1" />
 			</Target>
 			<Target title="wxSmithPlot">
 				<Option platforms="Windows;" />
-				<Option output="../../../devel/share/codeblocks/plugins/wxSmithPlot" prefix_auto="0" extension_auto="1" />
-				<Option working_dir="../../../devel" />
-				<Option object_output="../../../.objs/plugins/contrib/wxSmithPlot" />
+				<Option output="$(CB_BIN)/share/codeblocks/plugins/wxSmithPlot" prefix_auto="0" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)/plugins/contrib/wxSmithPlot" />
 				<Option type="3" />
 				<Option compiler="gcc" />
 				<Option parameters="--debug-log -na -ns -nd -p debug" />
@@ -42,14 +42,18 @@
 					<Add library="wxsmithlib" />
 					<Add library="codeblocks" />
 					<Add library="wxpropgrid" />
-					<Add directory="../../../devel" />
+					<Add directory="$(CB_LIB)/plugins/contrib/wxSmithPlot" />
+					<Add directory="$(CB_BIN)" />
 				</Linker>
 				<ExtraCommands>
-					<Add after="zip -j9 ..\..\..\devel\share\CodeBlocks\wxSmithPlot.zip manifest.xml" />
+					<Add after="zip -j9 $(CB_BIN)\share\CodeBlocks\wxSmithPlot.zip manifest.xml" />
 					<Mode after="always" />
 				</ExtraCommands>
 			</Target>
 			<Environment>
+				<Variable name="CB_BIN" value="..\..\..\devel" />
+				<Variable name="CB_LIB" value="..\..\..\lib" />
+				<Variable name="CB_OBJ" value="..\..\..\.objs" />
 				<Variable name="WX_CFG" value="" />
 				<Variable name="WX_SUFFIX" value="u" />
 			</Environment>
diff --git a/src/tools/cb_share_config/cb_share_config.cbp b/src/tools/cb_share_config/cb_share_config.cbp
index 44881f1..2abe629 100644
--- a/src/tools/cb_share_config/cb_share_config.cbp
+++ b/src/tools/cb_share_config/cb_share_config.cbp
@@ -7,13 +7,16 @@
 		<Option compiler="gcc" />
 		<Build>
 			<Target title="default">
-				<Option output="../../devel/cb_share_config" prefix_auto="0" extension_auto="1" />
-				<Option working_dir="../../../devel" />
-				<Option object_output="../../.objs/tools/cb_share_config" />
+				<Option output="$(CB_BIN)/cb_share_config" prefix_auto="0" extension_auto="1" />
+				<Option working_dir="$(CB_BIN)" />
+				<Option object_output="$(CB_OBJ)/tools/cb_share_config" />
 				<Option type="0" />
 				<Option compiler="gcc" />
 			</Target>
 			<Environment>
+				<Variable name="CB_BIN" value="..\..\devel" />
+				<Variable name="CB_LIB" value="..\..\lib" />
+				<Variable name="CB_OBJ" value="..\..\.objs" />
 				<Variable name="WX_CFG" value="" />
 				<Variable name="WX_SUFFIX" value="u" />
 			</Environment>
@@ -46,7 +49,7 @@
 			<Add library="txml" />
 			<Add library="wxmsw28$(WX_SUFFIX)" />
 			<Add directory="$(#WX.lib)/gcc_dll$(WX_CFG)" />
-			<Add directory="../../base/tinyxml" />
+			<Add directory="$(CB_LIB)" />
 		</Linker>
 		<Unit filename="app.cpp" />
 		<Unit filename="app.h" />
-- 
1.9.5.msysgit.1

